float64[] origin
float64[] normal
float64 quality